### PR TITLE
BAH - 1439 | Dockerise Bahmni ABDM Feature

### PR DESCRIPTION
**Pull request by SanoferSameera**

### CODE REVIEW

File: bahmni-docker/.env [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+DATA_PUSH_URL="http://ec2-65-0-184-72.ap-south-1.compute.amazonaws.com/hiu-api/data/notification"
+IDENTITY_JWK_URL= https://dev.ndhm.gov.in/auth/realms/consent-manager/protocol/openid-connect/certs
+HIU_ID="Bahmni"
+HIU_NAME="Bahmni"
+HIP_SECRET=
+HIP_CLIENTID=
+GATEWAY_URL="https://dev.abdm.gov.in/gateway"
+OPENMRS_URL="http://openmrs:8080/openmrs"
+OPENMRS_PASSWORD="Admin123"
+CORS_ALLOWED_ORIGIN="http://ec2-65-0-184-72.ap-south-1.compute.amazonaws.com"
```
Comment (swatigogia2020) [inline bahmni-docker/.env:81]: We can remove the env to be checked in, we can place the value in while running the image

File: bahmni-docker/.env [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+HIP_URL="http://ec2-65-0-184-72.ap-south-1.compute.amazonaws.com:8000"
```
Comment (swatigogia2020) [inline bahmni-docker/.env:82]: same as above

**Review by swatigogia2020 — CHANGES_REQUESTED** [review]

File: bahmni-docker/.env (hunk 1 above)
Comment (SanoferSameera) [reply bahmni-docker/.env]: removed

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    volumes:
+      - postgresql:/var/lib/postgresql
+      - postgresql_data:/var/lib/postgresql/data
+  rabbitmq:
+      image: "rabbitmq:3"
+      container_name: rabbitmq
+      ports:
+        - "5672:5672"
+  otp:
+      image: "ganesan92/otp-service"
```
Comment (gsluthra) [inline docker-compose.yml:173]: Whose account is otp-service being picked from?

File: .env [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+#Bahmni Web Environment Variables
+BAHMNI_WEB_IMAGE_TAG=latest
+BAHMNI_APPS_PATH=
+
+#Bahmni Proxy Environment Variables
+PROXY_IMAGE_TAG=latest
+
+POSTGRES_USER=postgres
+POSTGRES_PASSWORD=welcome
+RABBITMQ_USERNAME=guest
```
Comment (gsluthra) [inline .env:63]: I didn't realise RabbitMQ is also in the architecture for ABDM. This diagram here doesn't mention this: 
https://bahmni.atlassian.net/wiki/spaces/BAH/pages/2901114904/Bahmni+as+Health+Information+Provider+ABDM+NDHM

The above diagram also doesn't mention the OTP service.

**Comment by gsluthra** [comment]
Please provide description of major work done as part of this PR. The JIRA ticket also doesn't mention much info. 
https://bahmni.atlassian.net/browse/BAH-1439

File: docker-compose.yml (hunk 3 above)
Comment (swatigogia2020) [reply docker-compose.yml:173]: It came as part of projecteka fork

File: .env (hunk 4 above)
Comment (swatigogia2020) [reply .env:63]: I ll update the diagram gurpreet.

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    #volumes:
+      #- "${BAHMNI_APPS_PATH:?}/ui/app/:/usr/local/apache2/htdocs/bahmni"
+      #- "${BAHMNI_APPS_PATH:?}/ui/node_modules/@bower_components/:/usr/local/apache2/htdocs/bahmni/components"
+      #- "${BAHMNI_CONFIG_PATH:?}/:/usr/local/apache2/htdocs/bahmni_config/"
+
+  db:
+    container_name: db
+    image: "postgres:latest"
+    environment:
+      - POSTGRES_USER=postgres
```
Comment (gsluthra) [inline docker-compose.yml:67]: Should be read from .env file. This is hardcoded here.

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+      depends_on:
+        - db
+      environment:
+        whitelisted:numbers: ""
+
+  hiu-db-setup:
+      image: projecteka/hiu-db-initializer
+      container_name: hiu-db-setup
+      environment:
+        - 'JAVA_TOOL_OPTIONS=-Djdbc.url=jdbc:postgresql://db:5432/health_information_user -Djdbc.username=postgres -Djdbc.password=welcome'
```
Comment (gsluthra) [inline docker-compose.yml:91]: Should be read from .env file. This is hardcoded here. (port, credentials)

File: docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+      CONSENT_NOTIFICATION_URL: ${CONSENT_NOTIFICATION_URL:?}
+      CONSENT_MANAGEMENT_SUFFIX:  ${CONSENT_MANAGEMENT_SUFFIX:?}
+      CONSENT_MANAGEMENT_URL: ${CONSENT_MANAGEMENT_URL:?}
+      DEFAULT_PAGE_SIZE: 20
+      MAX_PAGE_SIZE: 100
+      LOCAL_STORAGE_PATH: /tmp/
+      HIU_CLIENT_ID: ${HIU_CLIENT_ID:?}
+      HIU_CLIENT_SECRET: ${HIU_CLIENT_SECRET:?}
+      POSTGRES_HOST: db
+      POSTGRES_PORT: 5432
```
Comment (gsluthra) [inline docker-compose.yml:137]: this should also be in .env file?

File: docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+      CONSENT_MANAGER_DB_NAME: health_information_user
+      POSTGRES_USER: postgres
```
Comment (gsluthra) [inline docker-compose.yml:139]: this should also be in .env file?

File: docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    volumes:
+      - postgresql:/var/lib/postgresql
+      - postgresql_data:/var/lib/postgresql/data
+  rabbitmq:
+      image: "rabbitmq:3"
+      container_name: rabbitmq
+      ports:
+        - "5672:5672"
+  otp:
+      image: "bahmniindiadistro/otp-service"
```
Comment (gsluthra) [inline docker-compose.yml:80]: The image tag is not needed (latest) -- to be picked from .env file?

File: docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+      container_name: otp
+      depends_on:
+        - db
+      environment:
+        whitelisted:numbers: ""
+
+  hiu-db-setup:
+      image: projecteka/hiu-db-initializer
```
Comment (gsluthra) [inline docker-compose.yml:88]: The image tag is not needed (latest) -- to be picked from .env file?

File: docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+      container_name: hiu-db-setup
+      environment:
+        - 'JAVA_TOOL_OPTIONS=-Djdbc.url=jdbc:postgresql://db:5432/health_information_user -Djdbc.username=postgres -Djdbc.password=welcome'
+
+  hip:
+    container_name: hip
+    image: bahmniindiadistro/hip
```
Comment (gsluthra) [inline docker-compose.yml:95]: The image tag is not needed (latest) -- to be picked from .env file?

File: docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+      Gateway:cmSuffix: ${CONSENT_MANAGEMENT_SUFFIX:?}
+      OtpService:BaseUrl: http://otp
+      hip:url: http://hip:8000
+      HEALTH_CHECK_DURATION: 30000
+      Cors:allowedOrigins: ${CORS_ALLOWED_ORIGIN:?}
+      Hip-Service:Url: ${HIP_URL:?}
+      Bahmni:Id: ${HIP_ID:?}
+
+  hiu:
+    image: bahmniindiadistro/hiu
```
Comment (gsluthra) [inline docker-compose.yml:119]: The image tag is not needed (latest) -- to be picked from .env file?

File: docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+      RABBITMQ_PASSWORD: ${RABBITMQ_PASSWORD:?}
+      MAX_IN_MEMORY_SIZE: 500MB
+      DATA_FLOW_CONSENT_REQUEST_WAIT_TIME: 1
+      DATA_FLOW_CONSENT_ARTEFACT_WAIT_TIME: 1
+      DATA_FLOW_PART_WAIT_TIME: 1440
+    ports:
+      - "9053:8003"
+
+  hiu-ui:
+    image: bahmniindiadistro/hiu-ui
```
Comment (gsluthra) [inline docker-compose.yml:160]: The image tag is not needed (latest) -- to be picked from .env file?

Comment (kavitha-sundararajan) [reply docker-compose.yml]: Modified the image tags to get from .env file

File: docker-compose.yml (hunk 8 above)
Comment (kavitha-sundararajan) [reply docker-compose.yml]: Modified all the hardcoded values to pick up from .env file

File: docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+
+  bahmni-web:
+    image: bahmniindiadistro/bahmni-web-abdm:${BAHMNI_WEB_IMAGE_TAG:?}
+    profiles: ["default","openmrs"]
+    #volumes:
+      #- "${BAHMNI_APPS_PATH:?}/ui/app/:/usr/local/apache2/htdocs/bahmni"
+      #- "${BAHMNI_APPS_PATH:?}/ui/node_modules/@bower_components/:/usr/local/apache2/htdocs/bahmni/components"
+      #- "${BAHMNI_CONFIG_PATH:?}/:/usr/local/apache2/htdocs/bahmni_config/"
+
+  db:
```
Comment (gsluthra) [inline docker-compose.yml:63]: A lot of services seem to depend on the same "db" service. Is the DB shared between HIP, HIU and OTP? Do they create separate schemas within one postgres server, or are all of them connecting to same DB (and hence tables are all created in same schema)?

Comment (gsluthra) [reply docker-compose.yml:63]: Ideally each service should be having its own dedicated DB.. and not share data across services via a common DB. Is that happening here?

File: .env [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+GATEWAY_JWK_URL="https://dev.abdm.gov.in/gateway/v0.5/certs"
+GATEWAY_BASE_URL="https://dev.abdm.gov.in/gateway/v0.5"
+HIU_DB_INITIALIZER_IMAGE_TAG=latest
+CONSENT_MANAGEMENT_URL="https://dev.ndhm.gov.in/cm"
+CONSENT_MANAGEMENT_SUFFIX="@sbx"
+CONSENT_NOTIFICATION_URL="http://localhost:8003/consent/notification"
+DATA_PUSH_URL=
+IDENTITY_JWK_URL= https://dev.ndhm.gov.in/auth/realms/consent-manager/protocol/openid-connect/certs
+GATEWAY_URL="https://dev.abdm.gov.in/gateway"
+OPENMRS_URL="http://openmrs:8080/openmrs"
```
Comment (gsluthra) [inline .env:85]: Why doesn't this variable use the same variables that already exisit? OPENMRS_HOST, PORT, etc? (Line 5/6)

Comment (gsluthra) [reply .env:85]: All OPENMRS creds are on top of the .env file. Why is this one lower in the file?

File: docker-compose.yml (hunk 14 above)
Comment (swatigogia2020) [reply docker-compose.yml:63]: They have their separate schemas , own dedicated DB

File: .env (hunk 15 above)
Comment (SanoferSameera) [reply .env]: we have reused the same variable

**Review by gsluthra — APPROVED** [review]

**Pull request merged**